ocaml/gencert/selfcert.ml

```diff
@@ -0,0 +1,125 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module C = Cmdliner
+module Rsa = Mirage_crypto_pk.Rsa
+module UX = Xapi_stdext_unix.Unixext
+open Rresult (* introduces >>= >>| and R *)
+
+module D = Debug.Make (struct let name = "gencert_selfcert" end)
+
+(** initialize the random number generator at program startup when this
+module is loaded. *)
+let () = Mirage_crypto_rng_unix.initialize ()
+
+(** [write_cert] writes a PKCS12 file to [path]. The typical file
+ extension would be ".pem". It attempts to do that atomically by
+ writing to a temporary file in the same directory first and renaming
+ the file at the end *)
+let write_certs path pkcs12 =
+  let f () =
+    UX.atomic_write_to_file path 0o400 @@ fun fd ->
```

Comment:
I expected `0o600` to be necessary here? (`600=rw- --- ---`, and `400=r-- --- ---`)

Reply:
This allows to move the certificate file, but not overwrite it with an empty file, add certificates the end, etc.

It's a small barrier to protect against accidental writes more than anything.